### ISSUE TITLE
Don't run flashing commands on UI thread

### ISSUE THREAD

**Issue by Tiebe** [badabing2005/PixelFlasher#216]
You shouldn't run the fastboot flashing comamnds on the UI thread, since this causes the application to hang (unable to resize or move window, etc) while flashing. 

Instead, run this on a background thread, and keep the main thread for the UI, while forwarding the logs to the this thread.

**Comment by badabing2005** [comment]
Agreed 100% in principal, however aside from the fact that it adds complexity, one needs to worry about each and every background process function and accordingly manage the UI by disabling features that shouldn't be executed while the background process is still running.
Forwarding logs from background thread process to UI console in real time rather than after completion is yet another non-trivial challenge.
Another concern is, if you let the user interact with the UI while running a background process that produces logs, (by design PixelFlasher is highly verbose to help users with troubleshooting) you end up with interlaced logs, which makes it quite difficult to troubleshoot and trace down problems.

I spend a lot of time helping users (on xda forums and here) with their support files troubleshooting problems, the last thing I want when tracing events is interlaced events and logs, I understand I could create log categories and group them and what not, but I have no intentions on refactoring a lot of the code just so that the UI is not blocked while flashing, incidentally it is not totally blocked, there are Yield operations almost everywhere.

I have implemented thread based running background shell process, you can see [here](https://github.com/badabing2005/PixelFlasher/blob/f0d0e67ee4bcb86184a456df4b9bf6997141db11/runtime.py#L3885) and it is used for certain invocations, however using it for flashing proved to be highly unreliable and problematic.

As stated in the readme, this was my first wxpyton project, and when I started coding this, I didn't know much about it, although I learned a lot along the way, by no means I consider myself an expert in the subject matter, and in fact I would welcome pull requests that make the project better and simpler.

**Comment by Tiebe** [comment]
Ahh, very understandable. I'll see if I have some time on my hands sometime soon to check it out.

**Comment by badabing2005** [comment]
> Using an API like https://wiki.wxpython.org/CallAfter looks very promising.
> 
> The idea is:
> 
>     * Run command via background thread, pass it a callback to update ui/on complete/whatever
> 
>     * UI receives the callback and locks/unlocks/updates etc (this can be implemented for logging too, you'd have to make the logger also run callbacks on a new log entry)
> 
> 
> It does require a fair bit of work to implement (talking from experience with similar situations in kotlin) but isn't usually too difficult for the better UX you get as a result

This is already being used in PF.
See [here](https://github.com/badabing2005/PixelFlasher/blob/f1c992c1008335ab3fd01d2670e85c27c0960c76/Main.py#L115), [here](https://github.com/badabing2005/PixelFlasher/blob/f1c992c1008335ab3fd01d2670e85c27c0960c76/runtime.py#L294), and few other places in PIF manager.
However this does not address the points I raised above.

If I have to wait for the completion callback, it's the same as locking, which by the way currently it doesn't totally lock, it yields at many points to allow the UI to refresh, but not necessarily interact with, which is the objective.
If we allow interaction, then I'd have to handle every UI case, which is not something I would want to do.

Thanks for the suggestion.
Sorry I missed your response here and this sat unanswered.

**Comment by sinancetinkaya** [comment]
I used to use wxpython on the main thread along with worker threads that were thread-safe accessing GUI. But your code needs to be heavily modified to do that. I can create a psudo code as an example for you if you're interested.

**Comment by badabing2005** [comment]
it's not just about being thread safe, it's handling how to manage the UI why an operation is in progress.
Allow / not allowing certain interaction based on context requires a lot logic code.
Not only that, it make the code complex and more difficult to maintain.
As you can see from my posts above,

**Comment by badabing2005** [comment]
@androidacy-user 
I welcome pull requests that does all the right things without breaking anything.
